src/main/kotlin/service/impl/TradingServiceImpl.kt

```diff
@@ -0,0 +1,156 @@
+package service.impl
+
+import currency.CryptoCurrency
+import exception.*
+import exchange.Exchange
+import service.TradingService
+import transaction.SwapTransaction
+import transaction.TradeTransaction
+import transaction.Transaction
+import user.User
+import user.UserStatus
+import wallet.Wallet
+import java.math.BigDecimal
+import java.math.RoundingMode
+import java.time.Clock
+import java.time.LocalDateTime
+import java.util.*
+
+class TradingServiceImpl(val clock: Clock, val random: Random) : TradingService {
+
+    companion object {
+        val ALLOWED_TRADE_STATUSES = setOf(UserStatus.APPROVED)
+        val ZERO_BALANCE_MESSAGE = "Not enough currency: %s on the balance. Your balance: 0, needed: %s"
+        val NOT_ENOUGH_CURRENCY_EXCEPTION_MESSAGE = "Not enough currency: %s on the balance. Your balance: %s, needed: %s"
+        val INCORRECT_PASSPHRASE_EXCEPTION_MESSAGE = "Passphrase is incorrect"
+        val TRADING_IS_NOT_ALLOWED_EXCEPTION_MESSAGE = "Trading for user with id %s is not allowed"
+        val EXCHANGE_PAIR_IS_NOT_PRESENT = "Exchange pair %s to %s is not present"
+        val TRANSACTION_FAILED_MESSAGE = "Transaction failed"
+    }
```

Comment:
Some notes about companion object. It should be either in the top of the class (as you do) or at the bottom - this specifies project.

You can also make these strings 'const'